### ISSUE TITLE
Move to split / combine keywords for "mapping" subgraphs over inputs

### ISSUE THREAD

**Issue by effigies** [nipype/pydra#18]
In order to ease transition for existing nipype users, a minimal goal is to ensure that `MapNode` and iterables+`JoinNode` functionality persists, in that nodes that fan out from a mapping can be recombined in the same order to produce lists of output.

Discussion was wide-ranging and perhaps at times experiencing a significant gravitas shortfall, but if I can remember correctly, we had the following genuine proposals:

| Fan-out | Fan-in |
|--|--|
| `map` | `unmap` |
| `map` | `collate` |
| `map` | `isquash` (i.e. squash by index) |
| `map` | `zap` |
| `split` | `combine` |
| `splitapply` | `combine` |

I'm sure I'm missing some. I would suggest we flesh out serious suggestions, with one comment per pair, and then have a quick round of voting, where heart (❤️) indicates your preference, thumbs-up (👍) indicates you would be happy enough with it, and thumbs-down (👎) indicates you *strongly* disapprove and will argue against it if it looks like we're settling on that one.

This is a separate discussion from value-based grouping or non-concatenation-based reductions, although you may want to weigh these options in terms of how easy they'll make naming such operations.

Edit: Please add any additional options you would like to vote on.

**Comment by kaczmarj** [comment]
i'm in favor of `split` / `combine`

**Comment by effigies** [comment]
Vote for `map` / `unmap`.

**Comment by effigies** [comment]
Vote for `map` / `collate`.

**Comment by effigies** [comment]
Vote for `map` / `isquash`.

**Comment by effigies** [comment]
Vote for `splitapply` / `combine`.

**Comment by miykael** [comment]
Not sure if this already came up, but what about `scatter` / `pool`? Or also just one of those terms.

**Comment by satra** [comment]
also ping @tyarkoni

**Comment by effigies** [comment]
If you want to add a voting option, you can just add a "Vote for `x` / `y`." comment. That said, not sure people are going to come back after one round of voting...

**Comment by yarikoptic** [comment]
I don't see map/reduce being considered, on purpose?

**Comment by effigies** [comment]
https://github.com/nipype/pydra/issues/18#issuecomment-430028146

**Comment by Shotgunosine** [comment]
Vote for fan_out/fan_in, or maybe fout/fin
(since that's what is used as the column headers)

**Comment by djarecka** [comment]
@yarikoptic - I strongly oppose map/reduce, since we are not using the same convention as typically people use when say reduceByKey or groupBy KEy

**Comment by effigies** [comment]
Vote for `fout` / `fin`.

Split out from https://github.com/nipype/pydra/issues/18#issuecomment-430030758.

**Comment by tyarkoni** [comment]
Why stop there! Vote for`fo` / `fi`!

**Comment by djarecka** [comment]
can someone remind/explain me what is the origin of `fan_out` / `fan_in`, `fout` / `fin`, etc.

**Comment by effigies** [comment]
`fan_out = map`, `fan_in = unmap`.

**Comment by effigies** [comment]
Looks like there's pretty strong consensus on `split` / `combine`. `map` / `unmap` comes in second. Neither has a blocking vote, and all others do, so unless there's a last minute objection, we'll go with `split` / `combine`.

**Comment by chrisfilo** [comment]
Excellent. Thanks for pushing this forward.

**Comment by djarecka** [comment]
since `split` / `combine` and `map` / `unmap` are the winners, what do you think about `map` \ `combine`

We started the discussion because we didn't have a good name for `combine`\`unmap`, etc. Do we really want to remove `map`?

**Comment by adelavega** [comment]
I actually feel that `split` may be more intuitive to novices, especially since it maps onto R dplyr terminology (pun intended)

**Comment by effigies** [comment]
It would have been nice to propose that prior to voting. I guess as a post-hoc test, can people thumbs up/thumbs down on *this* comment to override the current?

:+1: = Move to `map` / `combine`
:-1: = Stick with `split` / `combine`

There were 11 unique votes on `split` / `combine` (two both thumbs-up'd and hearted), so if this gets fewer than 11 votes to move to `map`, I would say we should leave it in `split`.

Re-summoning, just to make sure everybody gets a notification: @satra @chrisfilo @leej3 @oesteban @adelavega @yarikoptic @miykael @dPys @akeshavan @anibalsolon @rmarkello @Shotgunosine @mgxd @kaczmarj @tyarkoni

**Comment by leej3** [comment]
Here's a nice discussion of it for R: `http://www.stat.wisc.edu/~yandell/R_for_data_sciences/curate/tidyverse.html`. 

While they describe specific aggregation functions like count, they also show "non-concatenation-based reduction" with summarize. 

This just leaves the tricky operation of unmapping with state inputs presented in: https://docs.google.com/presentation/d/1yCjiJJdSGr9mlemwvecKsHUORy0gHWZSH6I5TWSnKR8/edit#slide=id.g43f27dddaf_0_27

To me that seems to be equivalent to combining, grouping by B and summarizing. Perhaps I'm wrong. It's a useful operation to implement as it's own thing and name effectively.

That's my long winded way of saying I think it should be split/combine!

**Comment by djarecka** [comment]
@effigies -  I know, that should have suggested earlier, sorry for this. But I just can't get used to the idea that we're removing "map" completely, seems to me that this word was a "natural" choice for long time. 
But whatever people decide... who had this weird idea to have democratic votes?! ;-)

**Comment by oesteban** [comment]
So, when are the midterm elections?

I'm running with @yarikoptic for ``map`` / ``reduce``

**Comment by effigies** [comment]
(Recognizing that there's joking going on...)

It's not a reduce function, though. And honestly, what we're describing isn't a map function, either. Map applies a function to some set of objects, reduce iteratively applies a binary function to a set of objects to return a single element.

```Haskell
map :: ((a -> b), [a]) -> [b]
reduce :: ((a, a) -> a), [a]) -> a
```

To get `map` behavior, you `split` a list, `apply` a subgraph, and `combine` the results.

So I think I'm increasingly a fan of moving away from the `map` idiom altogether.

**Comment by djarecka** [comment]
we are not only missing an option to reduce groups using function like sum, max, etc. (at least for now), but also we are not grouping in a way Spark/Pandas does.

**Comment by effigies** [comment]
Well, isn't `reduce` just a `combine`-ing node that has a function that reduces a list to a single element of the same type? e.g., `mean` in:

```Python
square = Node(lambda x: x**2, split='x')
mean = Node(delayed(np.mean)(a=square.results.out), combine='square.x')
root = Node(delayed(np.sqrt)(x=mean.results.out))
```

But yes, we should also get to the grouping by value.

**Comment by djarecka** [comment]
@effigies - if you are expecting that `mean` would give you mean of squares, that would not be the same as `reduceByKey` in Spark IMO (I would treat `square.x` as a "key").

But yes, I was asked to implement the version you described.

**Comment by satra** [comment]
implemented via #29 #30 #45
